### PR TITLE
Resolve issue 43

### PR DESCRIPTION
**Pull request by mamiksik**
Please do not peer review this PR yet. I created it only to get the Travis to run build and to check coverage diff.

### CODE REVIEW

File: pyodata/v2/model.py [hunk 1]
```diff
@@ -28,6 +30,113 @@ def modlog():
...
+    def resolve(self, ekseption):
+        self._logger.warning(repr(ekseption))
+
+
+class PolicySilent(ErrorPolicy):
+    def resolve(self, ekseption):
+        pass
+
+
+class AutoNumber(enum.Enum):
```
Comment (filak-sap) [inline pyodata/v2/model.py]: https://docs.python.org/3/library/enum.html#using-automatic-values

File: pyodata/v2/model.py [hunk 2]
```diff
@@ -840,135 +955,165 @@ def from_etree(schema_nodes):
...
+                except (KeyError, AttributeError) as ex:
+                    config.err_policy(ParserError.ENTITY_TYPE).resolve(ex)
+                    etype = NullType(entity_type.get('Name'))
+
                 decl.add_entity_type(etype)
 
         # resolve types of properties
         for stype in itertools.chain(schema.entity_types, schema.complex_types):
-            if stype.kind == Typ.Kinds.Complex:
+            if not isinstance(stype, NullType) and stype.kind == Typ.Kinds.Complex:
```
Comment (filak-sap) [inline pyodata/v2/model.py]: Please, do not join unrelated expressions via logic operators.

File: pyodata/v2/model.py [hunk 3]
```diff
@@ -840,135 +955,165 @@ def from_etree(schema_nodes):
...
+                    config.err_policy(ParserError.ASSOCIATION).resolve(ex)
+                    decl.associations[assoc.name] = NullAssociation(assoc.name)
+                else:
+                    decl.associations[assoc.name] = assoc
 
         # resolve navigation properties
         for stype in schema.entity_types:
             # skip collections
-            if stype.is_collection:
+            if isinstance(stype, NullType) or stype.is_collection:
```
Comment (filak-sap) [inline pyodata/v2/model.py]: Please, do not join unrelated expressions with logical operators.

Comment (mamiksik) [reply pyodata/v2/model.py]: danone

File: pyodata/v2/model.py (hunk 2 above)
Comment (mamiksik) [reply pyodata/v2/model.py]: danone

File: pyodata/v2/model.py (hunk 1 above)
Comment (mamiksik) [reply pyodata/v2/model.py]: Haven't found that before. Cool, I will rewrite to it.

**Comment by phanak-sap** [comment]
> Please do not peer review this PR yet. I created it only to get the Travis to run build and to check coverage diff.

There is Draft Pull Requests feature for that purpose:
https://github.blog/2019-02-14-introducing-draft-pull-requests/

File: docs/usage/initialization.rst [hunk 4]
```diff
@@ -42,6 +42,41 @@ hosted on private urls such as *customEdmxUrl.com* and *customEdmUrl.com*:
...
 
+Config
+------
+
+Class config provides easy to use wrapper for all parser configuration. These are:
+    - XML namespaces
+    - Parser policies(how parser act in case of invalid XML tag). We now support three types of policies:
+        - Policy fatal - parser raises exception and terminates
+        - Policy warning - parser reports given problem, then execute fallback code(e. g. replace unknown type with NullType) and then continues normally.
+        - Policy Silent - Same as warning except that error message is thrown away.
```
Comment (filak-sap) [inline docs/usage/initialization.rst]: Capitalization?

File: docs/usage/initialization.rst [hunk 5]
```diff
@@ -42,6 +42,41 @@ hosted on private urls such as *customEdmxUrl.com* and *customEdmUrl.com*:
...
+
+Parser policies can be specified individually for each XML tag(See enum ParserError for more details). If no policy is specified for the tag default policy is used.
```
Comment (filak-sap) [inline docs/usage/initialization.rst]: What is the tag policy? I guess there is a missing comma between tag and default.

File: docs/usage/initialization.rst [hunk 6]
```diff
@@ -42,6 +42,41 @@ hosted on private urls such as *customEdmxUrl.com* and *customEdmUrl.com*:
 
     northwind = pyodata.Client(SERVICE_URL, requests.Session(), namespaces=namespaces)
 
+Config
```
Comment (filak-sap) [inline docs/usage/initialization.rst]: Please, move this section to the end of Initialization.

File: pyodata/v2/model.py [hunk 7]
```diff
@@ -28,6 +30,106 @@ def modlog():
...
+class PolicyWarning(ErrorPolicy):
+    def __init__(self):
+        logging.basicConfig(format='%(levelname)s: %(message)s')
+        self._logger = logging.getLogger()
+
+    def resolve(self, ekseption):
+        self._logger.warning('[%s] %s', ekseption.__class__.__name__, str(ekseption))
+
+
+class PolicySilent(ErrorPolicy):
```
Comment (filak-sap) [inline pyodata/v2/model.py]: Please, rename to `PolicyIgnore`

File: docs/usage/initialization.rst [hunk 8]
```diff
@@ -101,3 +79,60 @@ Python client initialization:
...
+
+Dealing with errors during parsing metadata
+-------------------------------------------
+
+Class config provides easy to use wrapper for all parser configuration. These are:
+    - XML namespaces
+    - Parser policies(how parser act in case of invalid XML tag). We now support three types of policies:
+        - Policy fatal - parser raises exception and terminates
+        - Policy warning - parser reports given problem, then execute fallback code(e. g. replace unknown type with NullType) and then continues normally.
+        - Policy ignore - Same as warning except that error message is thrown away.
```
Comment (filak-sap) [inline docs/usage/initialization.rst]: Capitalization?

File: docs/usage/initialization.rst [hunk 9]
```diff
@@ -101,3 +79,60 @@ Python client initialization:
...
+
+Parser policies can be specified individually for each XML tag(See enum ParserError for more details). If no policy is specified for the tag, the default policy is used.
```
Comment (filak-sap) [inline docs/usage/initialization.rst]: s/tag(/tag (/

File: docs/usage/initialization.rst [hunk 10]
```diff
@@ -101,3 +79,60 @@ Python client initialization:
...
+
+.. code-block:: python
+
+    import pyodata
+    import requests
+
+    SERVICE_URL = 'http://services.odata.org/V2/Northwind/Northwind.svc/'
+
+    namespaces = {
+        'edmx': "customEdmxUrl.com",
```
Comment (filak-sap) [inline docs/usage/initialization.rst]: only '

File: docs/usage/initialization.rst [hunk 11]
```diff
@@ -101,3 +79,60 @@ Python client initialization:
...
+
+.. code-block:: python
+
+    import pyodata
+    import requests
+
+    SERVICE_URL = 'http://services.odata.org/V2/Northwind/Northwind.svc/'
+
+    namespaces = {
+        'edmx': "customEdmxUrl.com"
```
Comment (filak-sap) [inline docs/usage/initialization.rst]: only '

File: pyodata/v2/model.py [hunk 12]
```diff
@@ -28,6 +30,106 @@ def modlog():
...
+
+class Config:
+
+    def __init__(self,
+                 default_error_policy=None,
+                 custom_error_policies=None,
+                 xml_namespaces=None):
+
+        """
+        :param default_error_policy: ErrorPolicy (default PolicyFatal)
```
Comment (filak-sap) [inline pyodata/v2/model.py]: These two needs explanation.

File: docs/usage/initialization.rst (hunk 8 above)
Comment (filak-sap) [reply docs/usage/initialization.rst]: Periods?

File: pyodata/v2/model.py [hunk 13]
```diff
@@ -28,6 +30,106 @@ def modlog():
...
+    ENUM_TYPE = auto()
+    ENTITY_TYPE = auto()
+    COMPLEX_TYPE = auto()
+
+
+class Config:
+
+    def __init__(self,
+                 default_error_policy=None,
+                 custom_error_policies=None,
```
Comment (filak-sap) [inline pyodata/v2/model.py]: swap these two parameters

File: docs/usage/initialization.rst [hunk 14]
```diff
@@ -101,3 +79,60 @@ Python client initialization:
 
 
 For more information on client side SSL cerificationcas, please refer to this [gist](https://gist.github.com/mtigas/952344).
+
+Dealing with errors during parsing metadata
+-------------------------------------------
+
+Class config provides easy to use wrapper for all parser configuration. These are:
+    - XML namespaces
+    - Parser policies(how parser act in case of invalid XML tag). We now support three types of policies:
```
Comment (filak-sap) [inline docs/usage/initialization.rst]: s/policies(/policies (/

Comment (filak-sap) [reply docs/usage/initialization.rst]: s/We now/Now we/

File: docs/usage/initialization.rst (hunk 8 above)
Comment (filak-sap) [reply docs/usage/initialization.rst]: `Policy ignore - takes measures to recover from the error (e.g. if a type is referenced but now available, a stub type is used instead) without any further actions`

File: docs/usage/initialization.rst [hunk 15]
```diff
@@ -101,3 +79,60 @@ Python client initialization:
...
+        - Policy fatal - parser raises exception and terminates
+        - Policy warning - parser reports given problem, then execute fallback code(e. g. replace unknown type with NullType) and then continues normally.
```
Comment (filak-sap) [inline docs/usage/initialization.rst]: `Policy warning - the policy reports given problem, then executes a fallback code(e. g. replace unknown type with NullType) and then continues normally

File: docs/usage/initialization.rst [hunk 16]
```diff
@@ -101,3 +79,60 @@ Python client initialization:
 
 
 For more information on client side SSL cerificationcas, please refer to this [gist](https://gist.github.com/mtigas/952344).
+
+Dealing with errors during parsing metadata
+-------------------------------------------
+
```
Comment (filak-sap) [inline docs/usage/initialization.rst]: In the case where you need to consume a service which has not fully valid metadata document and is  not under your control, you can configure the metadata parser to try to recover from detected problems.

Parser recovery measures include actions such as using a stub entity type, if the parser cannot find a referenced entity type. The stub entity type allows the parser to continue processing the given metadata but causes fatal errors when accessed from the client.

File: docs/usage/initialization.rst [hunk 17]
```diff
@@ -101,3 +79,60 @@ Python client initialization:
...
+        xml_namespaces=namespaces,
+        default_error_policy=PolicyFatal(),
+        custom_error_policies={
+             ParserError.ANNOTATION: PolicyWarning(),
+             ParserError.ASSOCIATION: PolicyIgnore()
+        })
+
+    northwind = pyodata.Client(SERVICE_URL, requests.Session(), config=custom_config)
+
+Set custom namespaces(Deprecated - use config instead)
```
Comment (filak-sap) [inline docs/usage/initialization.rst]: s/namespaces(/namespaces (/

**Comment by filak-sap** [comment]
Thank you!

**Pull request merged**